File: TypeScript/src/14-Between_Two_Sets.ts
/**
 * Created on Thu Feb 23 2023
 * 
 * @author Carlos Páez
 */
'use strict'

import { WriteStream, createWriteStream } from "fs"

process.stdin.resume()
process.stdin.setEncoding( 'utf-8' )

let inputString: string = ''
let inputLines: string[] = []
let currentLine: number = 0

/* Reading the input from the console. */
process.stdin.on( 'data', function ( inputStdin: string ): void {
    inputString += inputStdin
} )

/* Reading the input from the console. */
process.stdin.on( 'end', function (): void {
    inputLines = inputString.split( '\n' )
    inputString = ''

    main()
} )

/**
 * It reads a line from the input file and returns it
 * @returns The current line of the inputLines array.
 */
function readLine (): string {
    return inputLines[ currentLine++ ]
}

/*
 * Complete the 'getTotalX' function below.
 *
 * The function is expected to return an INTEGER.
 * The function accepts following parameters:
 *  1. INTEGER_ARRAY a
 *  2. INTEGER_ARRAY b
 */

/**
 * It takes two arrays of integers, and returns the number of integers between the maximum value of the
 * first array and the minimum value of the second array that are evenly divisible by all the values in
 * the first array and evenly divisible by all the values in the second array.
 * @param {number[]} a - an array of integers
 * @param {number[]} b - an array of integers
 * @returns The number of integers that are between the two given arrays.
 */
function getTotalX ( a: number[], b: number[] ): number {
    if ( a.length < 1 || a.length > 10 || b.length < 1 || b.length > 10 ) {
        throw new Error( 'El tamaño de los arreglos debe estar entre 1 y 10.' )
    }

    const maxVal: number = 100
    if ( a.some( ( val ) => val < 1 || val > maxVal ) || b.some( ( val ) => val < 1 || val > maxVal ) ) {
        throw new Error( `Los valores de los arreglos deben estar entre 1 y ${ maxVal }.` )
    }

    const maxA: number = Math.max( ...a )
    const minB: number = Math.min( ...b )

    let count: number = 0

    for ( let i = maxA; i <= minB; i += maxA ) {
        if ( b.every( num => num % i === 0 ) && a.every( num => i % num === 0 ) ) {
            count++
        }
    }

    return count
}

function main () {
    const ws: WriteStream = createWriteStream( process.env[ 'OUTPUT_PATH' ] )

    const firstMultipleInput: string[] = readLine().replace( /\s+$/g, '' ).split( ' ' )

    const n: number = parseInt( firstMultipleInput[ 0 ], 10 )

    const m: number = parseInt( firstMultipleInput[ 1 ], 10 )

    const arr: number[] = readLine().replace( /\s+$/g, '' ).split( ' ' ).map( arrTemp => parseInt( arrTemp, 10 ) )

    const brr: number[] = readLine().replace( /\s+$/g, '' ).split( ' ' ).map( brrTemp => parseInt( brrTemp, 10 ) )

    const total: number = getTotalX( arr, brr )

    ws.write( total + '\n' )

    ws.end()
}

File: Python/14-Between_Two_Sets.py
# -*- coding: utf-8 -*-
""" 
Created on Thu Feb 23 2023

@author Carlos Páez
"""
from math import gcd
from functools import reduce


#
# Complete the 'getTotalX' function below.
#
# The function is expected to return an INTEGER.
# The function accepts following parameters:
#  1. INTEGER_ARRAY a
#  2. INTEGER_ARRAY b
#

def getTotalX(a, b):
    """
    We find the least common multiple of the first array, and the greatest common divisor of the second
    array. Then we count the number of multiples of the least common multiple that evenly divide the
    greatest common divisor

    :param a: an array of integers
    :param b: [2, 4]
    :return: The number of integers that are between two sets.
    """
    lest_common_multiple_a = reduce(lambda x, y: x * y // gcd(x, y), a)

    greatest_common_divisor_b = reduce(gcd, b)

    return sum(
        1 for x in range(lest_common_multiple_a, greatest_common_divisor_b + 1, lest_common_multiple_a)
        if greatest_common_divisor_b % x == 0
    )


if __name__ == '__main__':
    fptr = open('./output.txt', 'w')

    first_multiple_input = input().rstrip().split()

    n = int(first_multiple_input[0])
    m = int(first_multiple_input[1])

    arr = list(map(int, input().rstrip().split()))
    brr = list(map(int, input().rstrip().split()))

    total = getTotalX(arr, brr)

    fptr.write(str(total) + '\n')

    fptr.close()
